### DISCUSSION TITLE
Multi-manager with election

### DISCUSSION THREAD

**Discussion by grigorescu** [zeek/zeek#1400] [Projects]
Building off of the work in cluster/pools.zeek, the manager is the only cluster member which can't automatically failover. Given the unique nature of the manager (a single, authoritative member), this makes sense. However, operationally, I would like to have multiple managers, and have the Zeek cluster handle electing and promoting an "active" manager, and the rest being "standby" managers.

Data from peers to the managers go to all the managers (active and standby), and they maintain their individual local states. Responses from the manager back to the peers, or actions taken by the manager that should only occur once only occur on the "active" manager.

On cluster startup, the first manager in the list becomes active. There could be an event, which would allow the other managers to determine the active one. Those standby managers could handle `peer_removed`. If they see a standby manager removed, it gets removed from their list. If they see the active one removed, the next manager in the list promotes itself to active.

That's a pretty poor consensus algorithm, and I'm sure there will be some cases that fall through the cracks. Generally, though, I think Broker does a pretty good job of tracking both sides of a connection and detecting when it fails. Also, even this version is better than the current behavior, where losing a manager tends to lead to a cluster crash, as messages get queued up.

I think most of the work here is updating the scripts to differentiate between code that all managers should execute, and that which only the active one should execute.

**Comment by JustinAzoff** [comment]
IMHO, the manager just needs to die.  Have been working towards it for a few years now.  Logs were split off to the logger (and you can run more than one). Proxies are already redundant.  There's only a few things left that still go through the manager (notices, sumstats, intel,..?).  I say get everything the manager is doing able to run using proxies and then kill the manager entirely.  Might just need something similar to `publish_hrw` like `publish_leader`.  You can almost do this now by passing a constant to `publish_hrw` as the hash key.  Only downside is that while failover would work, it would immediately fail back, which might not be desired.

**Comment by ckreibich** [comment]
I think this is a control- vs data plane discussion. Vlad's points seem mainly about the control plane and Justin's more about the data plane. Vlad, your points strike me as different from how we use workers (as a source of events/logs from disjunct packet flows), proxies (load balancing events to other nodes), and loggers (distributing output workload). It's more basic and I could easily see it apply to other node types too — consider a proxy that's really a "crunch worker" that builds up expensive state you don't want to lose. So it might be good to back up and first define failover requirements, for example.

Managers seem tricky also because they have external implications: since they process external input (think ingestion of config options or intel data), local standby processes are one thing, switching to another machine another.

**Reply by grigorescu** [reply]
That's an interesting way of splitting it up; I hadn't thought about it like that. I think there are two separate issues:

1. Today, when the manager crashes and then restarts, the cluster usually cannot recover, at least not when it's processing any real workload. In my experience this exhibits as having the workers crash.
2. If the manager restarts, some data is lost. This could perhaps be solved with using the SQLite datastore rather than the in-memory one, but I'm not sure that all the state is captured by Broker datastores.

For now, I'm focusing on the first part -- how do we make the cluster more resilient by not crashing when a single node goes down? The second part becomes: how do we minimize the impact of a node restarting, by maintaining as much data/state as possible?

> Managers seem tricky also because they have external implications: since they process external input (think ingestion of config options or intel data), local standby processes are one thing, switching to another machine another.

Well, yes and no. I'd actually consider workers to be the trickiest, because there are physical network connections that force a particular architecture. Then, if there are differences in the system specs, that might limit which systems can run loggers or workers. Everything above layer 1 can be solved with software, though.

I'd argue that the examples you provided are a more specific case of what I mentioned -- the manager's main role is being a single, authoritative node. The reason that intel and config data is pushed out by the manager isn't due to some intrinsic limitation of that data only being able to reside on one system. The (TBD) orchestration that handles managing a multi-system cluster could easily copy that data around, just as `zeekctl` would copy `bin/zeek` around (note: just because it could, I don't think that it should actually do this). 

The reason we have that data pushed out by the manager is because we want a single, authoritative copy of it. If all the proxies would push that data out, we'd end up with conflicts if their versions wouldn't match, and no straightforward resolution path. I think I agree with Justin's comments, and would support moving  the manager's role to the proxies, and then having the concept of a single proxy "leader." Then, that proxy leader would be in charge of pushing out intel and config data, and it's up to the service owner to ensure that the data lives in the right place.

There is a trade-off in that while you do get resiliency, you have to copy data and updates to more systems. Maybe some users don't want that trade-off, and what to keep the current model. We could easily do that as well. If the manager moves to proxies, we could still support having a "manager" node, but that would just be syntactic sugar for that node being the only eligible proxy leader. If you don't have a manager, it would just use all the proxies and elect a leader from among them.

It's possible that this idea would also be influenced by the Broker ALM work, which has datastore clones.
